### PR TITLE
Added option to include exchange rate multiplier during import

### PR DESCRIPTION
**Pull request by carkom**
I've added an option to include an exchange rate during manual csv imports. It's very simple, just multiplies each transaction by the exchange rate during import. This allows for the user to use foreign accounts and keep their "home" currency constant.

Link to issue #612

### CODE REVIEW

**Comment by MatissJanis** [comment]
👋 I like the idea, but there's a few things we should polish product-wise.

1. "exchange rate" IMO is not a great label here. I'd suggest using "amount multiplier"
2. when it is provided - the "amount" column in the table should be updated to show the new amount; I suggest keeping the same design we have in the table column. So you would see "100 -> 120" if multiplier is set to "1.2"
3. the "multiplier" field IMO better fits in the right-side under the other import options; by default it should be disabled and it can be abled by toggling a checkbox (so people don't accidentally "scroll" in the field thus accidentally adding multiplier)
4. new "options" heading should be removed; it's redundant

Screenshot 2023-02-24 at 18 52 19

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Can't manually set the label sadly, so submitting an empty review

**Comment by j-f1** [comment]
I’m gonna spend some time thinking about alternate layouts for the import dialog that allow for adding more options and will update in an issue.

**Comment by carkom** [comment]
> I’m gonna spend some time thinking about alternate layouts for the import dialog that allow for adding more options and will update in an issue.

Here's an option. Not completely future proof but it does make this version much tidier.

image

**Comment by carkom** [comment]
I've updated the naming and location and set default to hide the box. Entries into the text box update the transactions table. Have a look!

**Comment by j-f1** [comment]
A couple of thoughts:

- Can you use the `Input` component for visual consistency? We don’t use native inputs anywhere else in the app as far as I know
- This only appears on CSV imports, do you think it would make sense to add it to the structured (QFX/OFX/etc) imports too?

**Comment by MatissJanis** [comment]
👋 (sorry, not yet a code review)

The modal seems to be broken for me. Getting `NaN` in the amount column.

Demo data: 
[2020-12-sanitised.csv](https://github.com/actualbudget/actual/files/10844030/2020-12-sanitised.csv)


Screenshot 2023-02-27 at 20 30 21

**Comment by carkom** [comment]
> 👋 (sorry, not yet a code review)
> 
> The modal seems to be broken for me. Getting `NaN` in the amount column.
> 
> Demo data: [2020-12-sanitised.csv](https://github.com/actualbudget/actual/files/10844030/2020-12-sanitised.csv)
> 


Thanks for the file. It helped me find my parsing errors and fix them.

**Comment by MatissJanis** [comment]
Ok, from product perspective this looks good now and works as expected. Great job!

I'll try to do a code-review tomorrow. But would you mind taking a look at the lint issues? Running `yarn lint --fix` should auto-fix them for you.

File: packages/loot-design/src/components/modals/ImportTransactions.js [hunk 1]
```diff
@@ -880,6 +984,8 @@ export function ImportTransactions({
             Import {transactions.length} transactions
           </ButtonWithLoading>
         </View>
+
+        <View style={{ flex: 1 }} />
```
Comment (MatissJanis) [inline packages/loot-design/src/components/modals/ImportTransactions.js:988]: **💬 suggestion:** ‏IMO this element is unnecessary (removed it from the DOM and the layout was still the same).

File: packages/loot-design/src/components/modals/ImportTransactions.js [hunk 2]
```diff
@@ -218,13 +219,23 @@ function parseAmount(amount, mapper) {
...
 
-function parseAmountFields(trans, splitMode, flipAmount) {
+function parseAmountFields(trans, splitMode, flipAmount, multiplierAmount) {
+  const multiplier = (() => {
+    if (multiplierAmount === '') {
+      return parseFloat('1');
+    } else {
+      return parseFloat(multiplierAmount);
+    }
+  })();
```
Comment (MatissJanis) [inline packages/loot-design/src/components/modals/ImportTransactions.js:229]: **🥜 nitpick:** ‏
```suggestion
  const multiplier = parseFloat(multiplierAmount) || 1.0;
```

File: packages/loot-design/src/components/modals/ImportTransactions.js [hunk 3]
```diff
@@ -380,6 +394,22 @@ function DateFormatSelect({
...
+    <View
+      style={{
+        flex: 1,
+        flexDirection: 'row',
+        alignItems: 'center',
+        userSelect: 'none',
+      }}
+    >
+      <Checkbox id="add_multiplier" checked={value} onChange={onChange} />
+      <label htmlFor="add_multiplier">Add Multiplier</label>
```
Comment (MatissJanis) [inline packages/loot-design/src/components/modals/ImportTransactions.js:408]: **👏 praise:** ‏love the attention to detail with `id` and `htmlFor`!

File: packages/loot-design/src/components/modals/ImportTransactions.js [hunk 4]
```diff
@@ -508,6 +538,26 @@ function FieldMappings({ transactions, mappings, onChange, splitMode }) {
...
 }
 
+function MultipliersField({ multiplierCB, value, onChange }) {
+  const styl = (() => {
+    if (multiplierCB) {
+      return 'inherit';
+    } else {
+      return 'none';
+    }
+  })();
```
Comment (MatissJanis) [inline packages/loot-design/src/components/modals/ImportTransactions.js:548]: **🥜 nitpick:** ‏
```suggestion
  const styl = multiplierCB ? 'inherit' : 'none';
```

File: packages/loot-design/src/components/modals/ImportTransactions.js [hunk 5]
```diff
@@ -526,6 +577,7 @@ export function ImportTransactions({
   let [fieldMappings, setFieldMappings] = useState(null);
   let [splitMode, setSplitMode] = useState(false);
   let [flipAmount, setFlipAmount] = useState(false);
+  let [multiplier, setMultiplier] = useState(false);
```
Comment (MatissJanis) [inline packages/loot-design/src/components/modals/ImportTransactions.js:580]: **🥜 nitpick:** ‏naming could be improved

```suggestion
  let [multiplierEnabled, setMultiplierEnabled] = useState(false);
```

**Review by MatissJanis — COMMENTED** [review]
Couple small suggestions: let me know what you think.

But overall looks very good! Can't wait to release this 🎉

File: packages/loot-design/src/components/modals/ImportTransactions.js (hunk 1 above)
Comment (carkom) [reply packages/loot-design/src/components/modals/ImportTransactions.js]: Yep, cool. Removed it.

**Comment by carkom** [comment]
Changes are all good. Commited them and fixed some variable naming issues. Should be ready!

**Review by MatissJanis — APPROVED** [review]
Code looks good!

@j-f1 any additional feedback from you before merging?

Also: ignore the failing e2e test. It's currently flaky and we're trying to fix it.

**Review by j-f1 — APPROVED** [review]
Nice, LGTM! Thank you for contributing.

**Comment by carkom** [comment]
Thanks for all your help @MatissJanis!

**Pull request merged**